set/reset UserWindow Title Text

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
New functions
**setUserWindowTitle**(string windowname, string text)
**resetUserWindowTitle**(string windowname)

Geyser.UserWindows:
userwindow:**setTitle(string text)**
userwindow:**resetTitle()**

Variable to setTitle at creation (Geyser.UserWindows):
**titleText** = "my custom title" -- has no default value

#### Motivation for adding to Mudlet
Was pretty simple to implement and was a feature which was missing.
Also userwindows should be adjustable :)

#### Other info (issues closed, discussion etc)

Edru2 committed:
first commit

New functions
setUserWindowTitle(string windowname, string text)
resetUserWindowT(string windowname)

Geyser.UserWindows:
userwindow:setTitle(string text)
userwindow:resetTitle()

Variables to setTitle at creation (Geyser.UserWindows):
titleText = "my custom title" -- has no default value

vadi2 commented:
:+1: 

I also think we make the default titles better - they became very verbose and I don't think it looks really good. Perhaps this PR is also an effect that the default titles aren't all that great?

Edru2 commented:
> +1
> 
> I also think we make the default titles better - they became very verbose and I don't think it looks really good. Perhaps this PR is also an effect that the default titles aren't all that great?

Yes, that is one of the reasons :smiley:

vadi2 commented:
Would you mind creating a separate PR for that? I think going back to the userwindow name is enough. If people want to differentiate them between profiles, they can with this PR.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -3590,6 +3590,35 @@ int TLuaInterpreter::openUserWindow(lua_State* L)
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#setUserWindowTitle
+int TLuaInterpreter::setUserWindowTitle(lua_State* L)
+{
+    QString name = QString();
+    QString text = QString();
```
```suggestion
    QString name, text;
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -3590,6 +3590,35 @@ int TLuaInterpreter::openUserWindow(lua_State* L)
+    QString name = QString();
+    QString text = QString();
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "setUserWindowTitle:  bad argument #1 type (name as string expected, got %s!)", luaL_typename(L, 1));
```
```suggestion
        lua_pushfstring(L, "setUserWindowTitle: bad argument #1 type (name as string expected, got %s!)", luaL_typename(L, 1));
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -3590,6 +3590,35 @@ int TLuaInterpreter::openUserWindow(lua_State* L)
+    }
+
+    if (!lua_isstring(L, 2)) {
+        lua_pushfstring(L, "setUserWindowTitle:  bad argument #2 type (title text as string expected, got %s!)", luaL_typename(L, 2));
```
```suggestion
        lua_pushfstring(L, "setUserWindowTitle: bad argument #2 type (title text as string expected, got %s!)", luaL_typename(L, 2));
```

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserUserWindow.lua:
```diff
@@ -103,6 +113,10 @@ function Geyser.UserWindow:new(cons)
       me:resize(cons.width,cons.height)
   end
 
+  if me.titleText then
```
If you change the constructor and remove the titleText, the user window's title won't be reset, I think.

Edru2 committed:
suggestions

SlySven commented:
You need to also allow this for the mapper widget if it is floating and the floating/dockable toolbars (the `TToolbar` class not the `TEasyButtonBar` class)...

SlySven commented:
It is important, IMHO to not obviously allow for NO title text for a window (fair enough I would accept a cheat of just a space character) but I think the previous, some might say verbose, title does serve an important function - do a right click on one of the toolbars in the editor and you will see what happens *without* a title text...

In summary for both this and #3586 I would be okay if the previous full text was used if an empty string was supplied for the text - as that would retain the existing behaviour for existing packages/scripts. And it needs to be equally applicable to the mapper and toolbars when they are floating or docked...

Also if you make the empty text (or nil) second argument put the original text back you only have to provide one function, not two and prevent accidental hiding the title by accident - if you really want to do a blank title you have to think about it and be deliberate...

Edru2 commented:
> It is important, IMHO to not obviously allow for NO title text for a window (fair enough I would accept a cheat of just a space character) but I think the previous, some might say verbose, title does serve an important function - do a right click on one of the toolbars in the editor and you will see what happens _without_ a title text...
> 
> In summary for both this and #3586 I would be okay if the previous full text was used if an empty string was supplied for the text - as that would retain the existing behaviour for existing packages/scripts.
> 
> Also if you make the empty text (or nil) second argument put the original text back you only have to provide one function, not two and prevent accidental hiding the title by accident - if you really want to do a blank title you have to think about it and be deliberate...

I'm not fully sure if I understand what you say but it will already reset the title if provided an empty string. The only way to provide an empty title is to put a space as title and that would be deliberate in my opinion.

>  And it needs to be equally applicable to the mapper and toolbars when they are floating or docked...

That is a good idea and something for another PR I think.

SlySven committed:
Revise: improvements to the implimentation of setUserWindowTitle(...)

Move the core code to the main `TConsole` for the profile concerned rather
than doing it in the mudlet singletion. This is part of the policy I trying
to put into practice of moving functionality out of the central mudlet
class when it does not need to be done in that shared between all profiles
core.

Allow the second argument to the Lua API setWindowTitle to be ommitted for
the reset to default case as well as doing that for the empty string case.

Add detection and error reporting for the user trying to set the title on
a console that is NOT a user window (and which would not have a title bar).

Also add handling of an error condition that should not be possible in
practice but which code analysers might otherwise moan about.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Please consider the *improvements* I propose in https://github.com/Edru2/Mudlet/pull/6 ...

Edru2 committed:
Merge pull request #6 from SlySven/userwindow-setTitle

Revise: improvements to the implimentation of setUserWindowTitle(...)

Edru2 committed:
resolve conflicts

Edru2 committed:
Merge branch 'development' into userwindow-setTitle

SlySven reviewed (approved):
And that is a :+1:  from me too. Now to :shipit:
